# King County Housing Linear Regression Project


* Student name: Seyma Tas
* Student pace:  Full time
* Scheduled project review date/time: 05-04-2020
* Instructor name: Amber Yandow
* Blog post URL:


## OSEMN Approach
1 - Obtaining the data 2 - Scrubbing the data 3 - Exploring the data 4 - Modeling the data 5 - iNterpreting the results


![title](picture2.png)


## Questions:

1) What are the main factors to predict the price of a house?

2) How much does a middle income family should invest to buy a standard house?
   
   or What are the properties of a house you can buy in King County if your budget is $400,000???
    
3) How can a middle income family increase the price of their house by %20 by spending %5 of the house cost???

4) Basement
 

In [98]:
%autosave 10

Autosaving every 10 seconds


The following cell is to print all the interactive output without resorting to print. 

In [99]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Import the necessary packages

First, we’ll import all of our necessary packages.

In [100]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib
import seaborn as sns

from scipy import stats
import scipy.stats as stats
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.outliers_influence import variance_inflation_factor

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score

sns.set(style='darkgrid')

Using matplotlib backend: MacOSX


Pandas has an options system that lets you customize some aspects of its behaviour, display-related options being those the user is most likely to adjust. In the following cell we are setting options to display more raws and columns.

In [101]:
# ## Display more information 

# pd.set_option('display.max_rows', None)
# pd.set_option('display.max_columns', None)
# pd.set_option('display.width', None)
# pd.set_option('display.max_colwidth', -1)

The following code is to hide the warning boxes.

In [102]:
import warnings
warnings.filterwarnings('ignore')

## Import the data set

Next, we’ll import our data and see what we’re working with. 

In [103]:
df=pd.read_csv('kc_house_data.csv')

Lets see what the data looks like by functions head(), tail(), shape(), info() and describe().

In [104]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0         NaN   0.0  ...      7        1180            0.0   
1      7242     2.0         0.0   0.0  ...      7        2170          400.0   
2     10000     1.0         0.0   0.0  ...      6         770            0.0   
3      5000     1.0         0.0   0.0  ...      7        1050          910.0   
4      8080     1.0         0.0   0.0  ...      8        1680            0.0   

  yr_built  yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0     1955           0.0    98178  47.5112 -122.257           1340        5650  
1     1951        1991.0    98125  47.7210 -122.319           1690        7639  
2     1933           NaN    98028  47.7379 -122.233           2720        8062  
3     1965           0.0    98136  47.5208 -122.393           1360        5000  
4     1987           0.0    98074  47.6168 -122.045           1800        7503  

[5 rows x 21 columns]

## Information about the data

### Column names and descriptions

* **id** - unique identified for a house
* **date** - Date house was sold
* **price** - Price is prediction target
* **bedrooms** - Number of Bedrooms/House
* **bathrooms** - Number of bathrooms/bedrooms
* **sqft_living** - Square footage of the home
* **sqft_lot** - Square footage of the lot
* **floors** - Total floors (levels) in house
* **waterfront** - House which has a view to a waterfront
* **view** - If the house has a view or not
* **condition** - How good the condition is ( Overall )
* **grade** - overall grade given to the housing unit, based on King County grading system
* **sqft_above** - square footage of house apart from basement
* **sqft_basement** - square footage of the basement
* **yr_built** - Built Year
* **yr_renovated** - Year when house was renovated
* **zipcode** - zipcode of the house
* **lat** - Latitude coordinate
* **long** - Longitude coordinate
* **sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors
* **sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors


### Condition

More detailed information about condition from the [King County's website](https://blue.kingcounty.com/Assessor/eRealProperty/ResidentialGlossary.aspx?idx=viewall&Parcel=7960900070&AreaReport=http://www.KingCounty.gov/depts/Assessor/Reports/area-reports/2019/residential-northeast/033.aspx#BuildingGrade).

(Relative to age and grade. Coded 1-5.)

1 = Poor- Worn out. Repair and overhaul needed on painted surfaces, roofing, plumbing, heating and numerous functional inadequacies. Excessive deferred maintenance and abuse, limited value-in-use, approaching abandonment or major reconstruction; reuse or change in occupancy is imminent. Effective age is near the end of the scale regardless of the actual chronological age.

2 = Fair- Badly worn. Much repair needed. Many items need refinishing or overhauling, deferred maintenance obvious, inadequate building utility and systems all shortening the life expectancy and increasing the effective age.

3 = Average- Some evidence of deferred maintenance and normal obsolescence with age in that a few minor repairs are needed, along with some refinishing. All major components still functional and contributing toward an extended life expectancy. Effective age and utility is standard for like properties of its class and usage.

4 = Good- No obvious maintenance required but neither is everything new. Appearance and utility are above the standard and the overall effective age will be lower than the typical property.

5= Very Good- All items well maintained, many having been overhauled and repaired as they have shown signs of wear, increasing the life expectancy and lowering the effective age with little deterioration or obsolescence evident with a high degree of utility.




### Building Grade

More detailed information about grade from the [King County's website](https://blue.kingcounty.com/Assessor/eRealProperty/ResidentialGlossary.aspx?idx=viewall&Parcel=7960900070&AreaReport=http://www.KingCounty.gov/depts/Assessor/Reports/area-reports/2019/residential-northeast/033.aspx#BuildingGrade).

Represents the construction quality of improvements. Grades run from grade 1 to 13. Generally defined as:

1-3	= Falls short of minimum building standards. Normally cabin or inferior structure

4	= Generally older, low quality construction. Does not meet code.

5	= Low construction costs and workmanship. Small, simple design.

6	= Lowest grade currently meeting building code. Low quality materials and simple designs.

7	= Average grade of construction and design. Commonly seen in plats and older sub-divisions.

8	= Just above average in construction and design. Usually better materials in both the exterior and interior finish work.

9	= Better architectural design with extra interior and exterior design and quality.

10	= Homes of this quality generally have high quality features. Finish work is better and more design quality is seen in the floor plans. Generally have a larger square footage.

11	= Custom design and higher quality finish work with added amenities of solid woods, bathroom fixtures and more luxurious options.

12	= Custom design and excellent builders. All materials are of the highest quality and all conveniences are present.

13	= Generally custom designed and built. Mansion level. Large amount of highest quality cabinet work, wood trim, marble, entry ways etc.


### Drop unnecessary columns

Id column doesn't tell us anything so we drop this column. 
Date column tells us the date the house was sold. All the dates are between 2014-05-02 and 2015-05-27. So we don't need to calculate inflation. But the sale month tells us information about sale price. May must be the month many houses are sold and the prices must be higher in May. Lets take the month and drop the date.

In [105]:
df['date'] = pd.to_datetime(df['date'])

In [106]:
df['date'].min()

Timestamp('2014-05-02 00:00:00')

In [107]:
df['date'].max()

Timestamp('2015-05-27 00:00:00')

In [108]:
df['sale_month'] = df['date'].map(lambda x: x.month)

In [109]:
df=df.drop(['id','date'],axis=1)

# Scraping and cleaning data

In [110]:
df.shape

(21597, 20)

We have 20 columns and 21597 rows,our sample size is enough to predict the price. There won't be problems in test and train splits. 

In [111]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 20 columns):
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
sale_month       21597 non-null int64
dtypes: float64(8), int64(11), object(1)
memory usage: 3.3+ MB


According to the information of data; we have some missing values in waterfront, view and year_renovated columns. We need to deal with missing data. Data is clean. It is necessary to inspect sqft_basement column. It should be an integer not a string.

### Numerical Data Stored as Strings

In [112]:
df.sqft_basement.value_counts()

0.0       12826
?           454
600.0       217
500.0       209
700.0       208
          ...  
20.0          1
1852.0        1
2250.0        1
3260.0        1
2310.0        1
Name: sqft_basement, Length: 304, dtype: int64

sqft_basement column is string because of 454 '?'. First we replace '?' with '0'. Then we change the data type to float and we find the 'sqft_basement' column from 'sqft_living' and 'sqft_above' columns.

In [113]:
# Replace '?' with '0'.
df['sqft_basement']=df['sqft_basement'].map(lambda x:float(x.replace('?', '0')))


In [114]:
#Replace 0 with difference between sqft_above and sqft_living
df['sqft_basement'].replace(to_replace='0', value = abs(df['sqft_living']-df['sqft_above']), inplace=True)


There is a house with 33 bedrooms. But its price is $640,000. Number of bathrooms is 1.75. It is a 3 bedroom house. 

In [115]:
df[df.bedrooms==33]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
15856  640000.0        33       1.75         1620      6000     1.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
15856         0.0   0.0          5      7        1040          580.0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
15856      1947           0.0    98103  47.6878 -122.331           1330   

       sqft_lot15  sale_month  
15856        4700           6

In [116]:
# Change 33 bedrooms to 3 bedrooms.
df['bedrooms'].replace(to_replace=33, value =3, inplace=True)


In [117]:
df.describe().round(1)

price  bedrooms  bathrooms  sqft_living   sqft_lot   floors  \
count    21597.0   21597.0    21597.0      21597.0    21597.0  21597.0   
mean    540296.6       3.4        2.1       2080.3    15099.4      1.5   
std     367368.1       0.9        0.8        918.1    41412.6      0.5   
min      78000.0       1.0        0.5        370.0      520.0      1.0   
25%     322000.0       3.0        1.8       1430.0     5040.0      1.0   
50%     450000.0       3.0        2.2       1910.0     7618.0      1.5   
75%     645000.0       4.0        2.5       2550.0    10685.0      2.0   
max    7700000.0      11.0        8.0      13540.0  1651359.0      3.5   

       waterfront     view  condition    grade  sqft_above  sqft_basement  \
count     19221.0  21534.0    21597.0  21597.0     21597.0        21597.0   
mean          0.0      0.2        3.4      7.7      1788.6          285.7   
std           0.1      0.8        0.7      1.2       827.8          439.8   
min           0.0      0.0        1.0      3.0       370.0            0.0   
25%           0.0      0.0        3.0      7.0      1190.0            0.0   
50%           0.0      0.0        3.0      7.0      1560.0            0.0   
75%           0.0      0.0        4.0      8.0      2210.0          550.0   
max           1.0      4.0        5.0     13.0      9410.0         4820.0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
count   21597.0       17755.0  21597.0  21597.0  21597.0        21597.0   
mean     1971.0          83.6  98078.0     47.6   -122.2         1986.6   
std        29.4         399.9     53.5      0.1      0.1          685.2   
min      1900.0           0.0  98001.0     47.2   -122.5          399.0   
25%      1951.0           0.0  98033.0     47.5   -122.3         1490.0   
50%      1975.0           0.0  98065.0     47.6   -122.2         1840.0   
75%      1997.0           0.0  98118.0     47.7   -122.1         2360.0   
max      2015.0        2015.0  98199.0     47.8   -121.3         6210.0   

       sqft_lot15  sale_month  
count     21597.0     21597.0  
mean      12758.3         6.6  
std       27274.4         3.1  
min         651.0         1.0  
25%        5100.0         4.0  
50%        7620.0         6.0  
75%       10083.0         9.0  
max      871200.0        12.0

Most houses were build around 1975; sold for $450K; had square footage of living spaces around 1910. You can also note, that distributions of price and variables such as sqft_lot are skewed to the upper tail.

## Detecting and Dealing With Null Values

In [118]:
print(df.isna().sum())

price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2376
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
sale_month          0
dtype: int64


We need to deal with missing values in waterfront, view and year_renovated columns.

#### Waterfront and view columns

In [119]:
df['view'].value_counts()

0.0    19422
2.0      957
3.0      508
1.0      330
4.0      317
Name: view, dtype: int64

Most of the houses don't have view. Houses in Seattle might have both ocean, lake and mountain views.

In [120]:
len(df[(df.view==4)&(df.waterfront>0)])

123

Almost %38 of the houses with good views are waterfront homes. The others must have Mount Rainier view or downtown view.

We are trying to decide what to do with missing values in waterfront column.

In [121]:
df['view'].isna().sum()

63

In [122]:
feats = ['view', 'waterfront']
df[feats].corr()

view  waterfront
view        1.000000    0.406654
waterfront  0.406654    1.000000

In [123]:
print('Number missing both:',
      len(df[(df.view.isnull())
      & (df.waterfront.isnull())]))

Number missing both: 6


In [124]:
len(df[(df.view==0)&(df.waterfront>0)])

0

In [125]:
len(df[(df.view==2)&(df.waterfront.isna())])

121

In [126]:
len(df[(df.view==3)&(df.waterfront>0)])

14

Waterfront and view columns are corrolated. If there was a view or waterfront is would going to be written. We can say missinf values in waterfront column can be filled with '0'.

In [127]:
df['view'].fillna(0,inplace=True)

In [128]:
df['waterfront'].value_counts()

0.0    19075
1.0      146
Name: waterfront, dtype: int64

In [129]:
df['waterfront'].nunique()

2

In [130]:
df['waterfront'].isna().sum()

2376

In [131]:
df['waterfront'].fillna(0,inplace=True)

In [132]:
feats = ['view', 'waterfront']
df[feats].corr()

view  waterfront
view        1.000000    0.380543
waterfront  0.380543    1.000000

#### Year_renovated column

In [133]:
df.yr_renovated.fillna(0, inplace=True)

## Check the distributions

 Let's check if the distributions of continous variables are normal. We can check if there is a high skewness or/and kurtosis.

In [134]:
df.hist(figsize=(20,15));


## Outliers

We have to be very careful about outliers. Waterfront houses are very expensive and they are detected as outliers. The real outlier in this data was 33 bedroom house. We have changed it to a 3 bedroom house. The others are real houses but some of them are very expensive because of some properties we don't have information. These expensive houses can have a pool or a special garden.They can be at walking distance to Grade A schools or a walking trail or playground. So we will get rid of %0.1 both from the bottom and from the top.

In [135]:
Q1 = df.quantile(0.001)
Q3 = df.quantile(0.999)
IQR = Q3 - Q1

df = df[~((df < (Q1 - 1.5 * IQR)) |(df > (Q3 + 1.5 * IQR))).any(axis=1)]

## Log transformation of data

Log transformation is needed to make the distributions much normal. Let's transform the continous variables by using logarithym of 'e'. 



In [136]:
non_normal = ['price','sqft_living', 'sqft_lot','sqft_above','sqft_basement',
              'yr_built', 'yr_renovated','sqft_living15', 'sqft_lot15']
for feat in non_normal:
    df[feat] = df[feat].map(lambda x: np.log1p(x))


## Normalizing the data

In [137]:
def norm_feat(series):
    return (series - series.mean())/series.std()
for feat in ['bathrooms', 'sqft_living', 'sqft_lot','sqft_above', 'sqft_basement', 'yr_built','yr_renovated','sqft_living15', 'sqft_lot15']:
    df[feat] = norm_feat(df[feat])


In [138]:
df.describe().round(2)

price  bedrooms  bathrooms  sqft_living  sqft_lot    floors  \
count  21594.00  21594.00   21594.00     21594.00  21594.00  21594.00   
mean      13.05      3.37      -0.00         0.00     -0.00      1.49   
std        0.53      0.90       1.00         1.00      1.00      0.54   
min       11.26      1.00      -2.10        -3.86     -3.04      1.00   
25%       12.68      3.00      -0.48        -0.67     -0.52      1.00   
50%       13.02      3.00       0.17         0.01     -0.06      1.50   
75%       13.38      4.00       0.50         0.69      0.32      2.00   
max       15.86     11.00       7.66         4.63      5.53      3.50   

       waterfront      view  condition     grade  sqft_above  sqft_basement  \
count    21594.00  21594.00   21594.00  21594.00    21594.00       21594.00   
mean         0.01      0.23       3.41      7.66        0.00          -0.00   
std          0.08      0.76       0.65      1.17        1.00           1.00   
min          0.00      0.00       1.00      3.00       -3.47          -0.79   
25%          0.00      0.00       3.00      7.00       -0.73          -0.79   
50%          0.00      0.00       3.00      7.00       -0.10          -0.79   
75%          0.00      0.00       4.00      8.00        0.72           1.21   
max          1.00      4.00       5.00     13.00        4.11           1.90   

       yr_built  yr_renovated   zipcode       lat      long  sqft_living15  \
count  21594.00      21594.00  21594.00  21594.00  21594.00       21594.00   
mean       0.00          0.00  98077.96     47.56   -122.21           0.00   
std        1.00          1.00     53.51      0.14      0.14           1.00   
min       -2.45         -0.19  98001.00     47.16   -122.52          -4.73   
25%       -0.67         -0.19  98033.00     47.47   -122.33          -0.71   
50%        0.14         -0.19  98065.00     47.57   -122.23          -0.07   
75%        0.88         -0.19  98118.00     47.68   -122.12           0.69   
max        1.48          5.30  98199.00     47.78   -121.32           3.65   

       sqft_lot15  sale_month  
count    21594.00    21594.00  
mean        -0.00        6.57  
std          1.00        3.11  
min         -3.06        1.00  
25%         -0.52        4.00  
50%         -0.03        6.00  
75%          0.32        9.00  
max          5.27       12.00

In [139]:
df.to_csv("king_county_dataset_cleaned.csv", index=False)


## Check the distributions again!

In [140]:
df.hist(figsize=(20,15));


We optained normal distributions in 'price', 'sqft_living','sqft_above', 'sqft_lot', 'yr_built', 'yr_renovated', 'sqft_living15' columns which is very good to apply linear regression.

## One-Hot Encoding Categorical Columns


#### Identifying Categorical Variables

These columns can be thought as categorical values. Bedrooms, floors, waterfront, wiev, grade and zipcode. We change their type to 'category' and make new columns by using pandas get_dummies() function.


In [141]:
df['bedrooms'] = df['bedrooms'].astype('category')
df['floors'] = df['floors'].astype('category')
df['waterfront'] = df['waterfront'].astype('category')
df['view'] = df['view'].astype('category')
df['condition'] = df['condition'].astype('category')
df['grade'] = df['grade'].astype('category')
df['zipcode'] = df['zipcode'].astype('category')

In [142]:
feats = ['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'sqft_living15', 'sqft_lot15']
df_feats = df[feats]
df_feats_zipcode=df_feats.copy()
df_feats = pd.get_dummies(df_feats, drop_first=True)

# Exploring the data

In [143]:
df.head()

price bedrooms  bathrooms  sqft_living  sqft_lot floors waterfront  \
0  12.309987        3  -1.451612    -1.125683 -0.388509    1.0        0.0   
1  13.195616        3   0.174831     0.709834 -0.112772    2.0        0.0   
2  12.100718        2  -1.451612    -2.131831  0.245671    1.0        0.0   
3  13.311331        4   1.150697     0.070792 -0.524259    1.0        0.0   
4  13.142168        3  -0.150458    -0.292723  0.008853    1.0        0.0   

  view condition grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0  0.0         3     7   -0.753610      -0.785202 -0.537420     -0.188896   
1  0.0         3     7    0.673014       1.112346 -0.674345      5.291902   
2  0.0         3     6   -1.752832      -0.785202 -1.294001     -0.188896   
3  0.0         5     7   -1.026876       1.372122 -0.196328     -0.188896   
4  0.0         3     8    0.073617      -0.785202  0.548008     -0.188896   

  zipcode      lat     long  sqft_living15  sqft_lot15  sale_month  
0   98178  47.5112 -122.257      -1.035409   -0.395548          10  
1   98125  47.7210 -122.319      -0.326807   -0.023625          12  
2   98028  47.7379 -122.233       1.126666    0.042836           2  
3   98136  47.5208 -122.393      -0.990173   -0.546252          12  
4   98074  47.6168 -122.045      -0.134239   -0.045777           2

In [144]:
df_feats.head()

price  bathrooms  sqft_living  sqft_lot  sqft_above  sqft_basement  \
0  12.309987  -1.451612    -1.125683 -0.388509   -0.753610      -0.785202   
1  13.195616   0.174831     0.709834 -0.112772    0.673014       1.112346   
2  12.100718  -1.451612    -2.131831  0.245671   -1.752832      -0.785202   
3  13.311331   1.150697     0.070792 -0.524259   -1.026876       1.372122   
4  13.142168  -0.150458    -0.292723  0.008853    0.073617      -0.785202   

   yr_built  yr_renovated  sqft_living15  sqft_lot15  ...  zipcode_98146  \
0 -0.537420     -0.188896      -1.035409   -0.395548  ...              0   
1 -0.674345      5.291902      -0.326807   -0.023625  ...              0   
2 -1.294001     -0.188896       1.126666    0.042836  ...              0   
3 -0.196328     -0.188896      -0.990173   -0.546252  ...              0   
4  0.548008     -0.188896      -0.134239   -0.045777  ...              0   

   zipcode_98148  zipcode_98155  zipcode_98166  zipcode_98168  zipcode_98177  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   zipcode_98178  zipcode_98188  zipcode_98198  zipcode_98199  
0              1              0              0              0  
1              0              0              0              0  
2              0              0              0              0  
3              0              0              0              0  
4              0              0              0              0  

[5 rows x 113 columns]

In [145]:
df_plot=df[['price', 'bedrooms','bathrooms','sqft_living', 'sqft_lot', 'sqft_above', 'sqft_basement', 'yr_built', 'lat','long','sqft_living15','sqft_lot15']].copy()

In [146]:
sns.pairplot(df_plot);

## Check for Linearity

jointplots

## Check for Multicollinearity

In [147]:
# df_target = df_no_outliers.loc['price']
df_heatmap=df.drop(['price','zipcode', 'lat', 'long'],axis=1)
plt.figure(figsize=(20,20))
ax=sns.heatmap(df_heatmap.corr(), center=0, linewidths=.5,annot=True);
bottom, top = ax.get_ylim()
ax.set_ylim(bottom + 0.5, top - 0.5)
plt.show()

<Figure size 2880x4000 with 0 Axes>

(10.0, 0.0)

In [148]:
df_heatmap.corr() > 0.75


bathrooms  sqft_living  sqft_lot  sqft_above  sqft_basement  \
bathrooms           True         True     False       False          False   
sqft_living         True         True     False        True          False   
sqft_lot           False        False      True       False          False   
sqft_above         False         True     False        True          False   
sqft_basement      False        False     False       False           True   
yr_built           False        False     False       False          False   
yr_renovated       False        False     False       False          False   
sqft_living15      False        False     False       False          False   
sqft_lot15         False        False      True       False          False   
sale_month         False        False     False       False          False   

               yr_built  yr_renovated  sqft_living15  sqft_lot15  sale_month  
bathrooms         False         False          False       False       False  
sqft_living       False         False          False       False       False  
sqft_lot          False         False          False        True       False  
sqft_above        False         False          False       False       False  
sqft_basement     False         False          False       False       False  
yr_built           True         False          False       False       False  
yr_renovated      False          True          False       False       False  
sqft_living15     False         False           True       False       False  
sqft_lot15        False         False          False        True       False  
sale_month        False         False          False       False        True

# Modeling the data

## Baseline Model 

In [149]:
outcome = 'price'
x_cols = [ 'bathrooms', 'sqft_living', 'sqft_lot', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'sqft_living15',
       'sqft_lot15', 'bedrooms_2', 'bedrooms_3', 'bedrooms_4', 'bedrooms_5',
       'bedrooms_6', 'bedrooms_7', 'bedrooms_8', 'bedrooms_9', 'bedrooms_10',
       'bedrooms_11', 'floors_1_5', 'floors_2_0', 'floors_2_5', 'floors_3_0',
       'floors_3_5', 'waterfront_1_0', 'view_1_0', 'view_2_0', 'view_3_0',
       'view_4_0', 'condition_2', 'condition_3', 'condition_4', 'condition_5',
       'grade_4', 'grade_5', 'grade_6', 'grade_7', 'grade_8', 'grade_9',
       'grade_10', 'grade_11', 'grade_12', 'grade_13', 'zipcode_98002',
       'zipcode_98003', 'zipcode_98004', 'zipcode_98005', 'zipcode_98006',
       'zipcode_98007', 'zipcode_98008', 'zipcode_98010', 'zipcode_98011', 
       'zipcode_98019', 'zipcode_98022', 'zipcode_98023', 'zipcode_98024',
       'zipcode_98027', 'zipcode_98028', 'zipcode_98029', 'zipcode_98030',
       'zipcode_98031', 'zipcode_98032', 'zipcode_98033', 'zipcode_98034',
       'zipcode_98038', 'zipcode_98039', 'zipcode_98040', 'zipcode_98042',
       'zipcode_98045', 'zipcode_98052', 'zipcode_98053', 'zipcode_98055',
       'zipcode_98056', 'zipcode_98058', 'zipcode_98059', 'zipcode_98065',
       'zipcode_98070', 'zipcode_98072', 'zipcode_98074', 'zipcode_98075',
       'zipcode_98077', 'zipcode_98092', 'zipcode_98102', 'zipcode_98103',
       'zipcode_98105', 'zipcode_98106', 'zipcode_98107', 'zipcode_98108',
       'zipcode_98109', 'zipcode_98112', 'zipcode_98115', 'zipcode_98116',
       'zipcode_98117', 'zipcode_98118', 'zipcode_98119', 'zipcode_98122',
       'zipcode_98125', 'zipcode_98126', 'zipcode_98133', 'zipcode_98136',
       'zipcode_98144', 'zipcode_98146', 'zipcode_98148', 'zipcode_98155',
       'zipcode_98166', 'zipcode_98168', 'zipcode_98177', 'zipcode_98178',
       'zipcode_98188', 'zipcode_98198', 'zipcode_98199','zipcode_98014']


In [150]:
# df_feats.columns = [col.replace('.', '_', inplace=True) for col in df_feats.columns]
# df_feats.columns = [col.replace(' ', '_', inplace=True) for col in df_feats.columns]
df_feats.columns = df_feats.columns.str.replace(r"[.]", "_")

In [151]:
df_feats.columns[:50]

Index(['price', 'bathrooms', 'sqft_living', 'sqft_lot', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'sqft_living15',
       'sqft_lot15', 'bedrooms_2', 'bedrooms_3', 'bedrooms_4', 'bedrooms_5',
       'bedrooms_6', 'bedrooms_7', 'bedrooms_8', 'bedrooms_9', 'bedrooms_10',
       'bedrooms_11', 'floors_1_5', 'floors_2_0', 'floors_2_5', 'floors_3_0',
       'floors_3_5', 'waterfront_1_0', 'view_1_0', 'view_2_0', 'view_3_0',
       'view_4_0', 'condition_2', 'condition_3', 'condition_4', 'condition_5',
       'grade_4', 'grade_5', 'grade_6', 'grade_7', 'grade_8', 'grade_9',
       'grade_10', 'grade_11', 'grade_12', 'grade_13', 'zipcode_98002',
       'zipcode_98003', 'zipcode_98004', 'zipcode_98005', 'zipcode_98006',
       'zipcode_98007'],
      dtype='object')

In [152]:
df_feats.columns[50:]

Index(['zipcode_98008', 'zipcode_98010', 'zipcode_98011', 'zipcode_98014',
       'zipcode_98019', 'zipcode_98022', 'zipcode_98023', 'zipcode_98024',
       'zipcode_98027', 'zipcode_98028', 'zipcode_98029', 'zipcode_98030',
       'zipcode_98031', 'zipcode_98032', 'zipcode_98033', 'zipcode_98034',
       'zipcode_98038', 'zipcode_98039', 'zipcode_98040', 'zipcode_98042',
       'zipcode_98045', 'zipcode_98052', 'zipcode_98053', 'zipcode_98055',
       'zipcode_98056', 'zipcode_98058', 'zipcode_98059', 'zipcode_98065',
       'zipcode_98070', 'zipcode_98072', 'zipcode_98074', 'zipcode_98075',
       'zipcode_98077', 'zipcode_98092', 'zipcode_98102', 'zipcode_98103',
       'zipcode_98105', 'zipcode_98106', 'zipcode_98107', 'zipcode_98108',
       'zipcode_98109', 'zipcode_98112', 'zipcode_98115', 'zipcode_98116',
       'zipcode_98117', 'zipcode_98118', 'zipcode_98119', 'zipcode_98122',
       'zipcode_98125', 'zipcode_98126', 'zipcode_98133', 'zipcode_98136',
       'zipcode_98144', '

In [153]:
# Fitting the actual model
predictors = '+'.join(x_cols)
formula = outcome + '~' + predictors
model = ols(formula=formula, data=df_feats).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.885
Model:                            OLS   Adj. R-squared:                  0.885
Method:                 Least Squares   F-statistic:                     1481.
Date:                Tue, 28 Apr 2020   Prob (F-statistic):               0.00
Time:                        17:00:32   Log-Likelihood:                 6599.3
No. Observations:               21594   AIC:                        -1.297e+04
Df Residuals:                   21481   BIC:                        -1.207e+04
Df Model:                         112                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         12.6470      0.183     69.203      0.000      12.289      13.005
bathrooms          0.0293      0.002     12.809      0.000       0.025       0.034
sqft_living        0.0994      0.006     16.597      0.000       0.088       0.111
sqft_lot           0.0642      0.003     19.750      0.000       0.058       0.071
sqft_above         0.0817      0.006     13.783      0.000       0.070       0.093
sqft_basement      0.0134      0.003      4.596      0.000       0.008       0.019
yr_built          -0.0031      0.002     -1.319      0.187      -0.008       0.001
yr_renovated       0.0125      0.001      9.550      0.000       0.010       0.015
sqft_living15      0.0485      0.002     22.091      0.000       0.044       0.053
sqft_lot15        -0.0171      0.003     -5.309      0.000      -0.023      -0.011
bedrooms_2        -0.0162      0.014     -1.188      0.235      -0.043       0.011
bedrooms_3        -0.0400      0.014     -2.903      0.004      -0.067      -0.013
bedrooms_4        -0.0425      0.014     -3.002      0.003      -0.070      -0.015
bedrooms_5        -0.0609      0.015     -4.070      0.000      -0.090      -0.032
bedrooms_6        -0.0933      0.018     -5.143      0.000      -0.129      -0.058
bedrooms_7        -0.2045      0.033     -6.249      0.000      -0.269      -0.140
bedrooms_8        -0.1224      0.052     -2.343      0.019      -0.225      -0.020
bedrooms_9        -0.2578      0.075     -3.430      0.001      -0.405      -0.111
bedrooms_10       -0.2774      0.105     -2.655      0.008      -0.482      -0.073
bedrooms_11        0.0432      0.180      0.240      0.810      -0.309       0.396
floors_1_5         0.0036      0.005      0.709      0.478      -0.006       0.014
floors_2_0        -0.0144      0.004     -3.329      0.001      -0.023      -0.006
floors_2_5         0.0048      0.015      0.320      0.749      -0.025       0.034
floors_3_0        -0.0824      0.010     -8.670      0.000      -0.101      -0.064
floors_3_5        -0.0605      0.068     -0.885      0.376      -0.194       0.074
waterfront_1_0     0.4416      0.018     23.938      0.000       0.405       0.478
view_1_0           0.1114      0.010     10.940      0.000       0.091       0.131
view_2_0           0.1021      0.006     16.409      0.000       0.090       0.114
view_3_0           0.1733      0.008     20.476      0.000       0.157       0.190
view_4_0           0.2911      0.013     22.742      0.000       0.266       0.316
condition_2        0.1550      0.036      4.288      0.000       0.084       0.226
condition_3        0.2881      0.034      8.561      0.000       0.222       0.354
condition_4        0.3283      0.034      9.750      0.000       0.262       0.394
condition_5        0.3945      0.034     11.642      0.000       0.328       0.461
grade_4     

In [154]:
fig = sm.graphics.qqplot(model.resid, dist=stats.norm, line='45', fit=True)

In [155]:
plt.scatter(model.predict(df_feats[x_cols]), model.resid)
plt.plot(model.predict(df_feats[x_cols]), [0 for i in range(len(df_feats))])
plt.show()

In [156]:
X = df_feats[x_cols]

In [157]:
vif = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
list(zip(x_cols, vif))

[('bathrooms', 3.530401994851968),
 ('sqft_living', 24.222067447269048),
 ('sqft_lot', 7.15187700835907),
 ('sqft_above', 23.759504798232825),
 ('sqft_basement', 5.759001916723862),
 ('yr_built', 3.669062352261675),
 ('yr_renovated', 1.156633366476584),
 ('sqft_living15', 3.2570041356625135),
 ('sqft_lot15', 7.010293489551642),
 ('bedrooms_2', 16.112985759119614),
 ('bedrooms_3', 58.32857026250645),
 ('bedrooms_4', 43.242884397407956),
 ('bedrooms_5', 11.207511327710838),
 ('bedrooms_6', 2.802938675485087),
 ('bedrooms_7', 1.2742432701163189),
 ('bedrooms_8', 1.1101049331222081),
 ('bedrooms_9', 1.0613189606687237),
 ('bedrooms_10', 1.0256945357111498),
 ('bedrooms_11', 1.0119550156020927),
 ('floors_1_5', 1.5410952155622177),
 ('floors_2_0', 4.799164455602561),
 ('floors_2_5', 1.1360435914970166),
 ('floors_3_0', 1.7287549653028464),
 ('floors_3_5', 1.0242600394851575),
 ('waterfront_1_0', 1.5553823568383076),
 ('view_1_0', 1.0721670412106308),
 ('view_2_0', 1.1580940140641796),
 ('vi

## K-fold Cross Validation for Model 1

In [158]:
X = df_feats[x_cols]
y = df_feats['price']

linreg = LinearRegression()
cvs_10=cross_val_score(linreg, X, y, cv=10)
cvs_10_res_mean = round(np.mean(cvs_10),5)
cv_10_res_std = round(np.std(cvs_10),5)
print(f'Mean cross validation score for 10 fold is {cvs_10_res_mean}')
print(f'Standard deviation of the score is {cv_10_res_std}')


Mean cross validation score for 10 fold is -3.401270363261404e+20
Standard deviation of the score is 1.0136709916109482e+21


## Perform stepwise selection

In [159]:
import statsmodels.api as sm

def stepwise_selection(X, y, 
                       initial_list=[], 
                       threshold_in=0.01, 
                       threshold_out = 0.05, 
                       verbose=True):
    """ Perform a forward-backward feature selection 
    based on p-value from statsmodels.api.OLS
    Arguments:
        X - pandas.DataFrame with candidate features
        y - list-like with the target
        initial_list - list of features to start with (column names of X)
        threshold_in - include a feature if its p-value < threshold_in
        threshold_out - exclude a feature if its p-value > threshold_out
        verbose - whether to print the sequence of inclusions and exclusions
    Returns: list of selected features 
    Always set threshold_in < threshold_out to avoid infinite looping.
    See https://en.wikipedia.org/wiki/Stepwise_regression for the details
    """
    included = list(initial_list)
    while True:
        changed=False
        # forward step
        excluded = list(set(X.columns)-set(included))
        new_pval = pd.Series(index=excluded)
        for new_column in excluded:
            model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included+[new_column]]))).fit()
            new_pval[new_column] = model.pvalues[new_column]
        best_pval = new_pval.min()
        if best_pval < threshold_in:
            best_feature = new_pval.idxmin()
            included.append(best_feature)
            changed=True
            if verbose:
                print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

        # backward step
        model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
        # use all coefs except intercept
        pvalues = model.pvalues.iloc[1:]
        worst_pval = pvalues.max() # null if pvalues is empty
        if worst_pval > threshold_out:
            changed=True
            worst_feature = pvalues.argmax()
            included.remove(worst_feature)
            if verbose:
                print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
        if not changed:
            break
    return included

In [160]:
X = df_feats[x_cols]
# X = df_feats.drop('price', axis=1)
y = df_feats['price']

result = stepwise_selection(X, y, verbose = False)
print('resulting features:')
print(result)

resulting features:
['grade_9', 'grade_7', 'sqft_living15', 'sqft_living', 'grade_6', 'zipcode_98004', 'zipcode_98023', 'grade_8', 'grade_5', 'waterfront_1_0', 'zipcode_98003', 'zipcode_98092', 'zipcode_98038', 'zipcode_98002', 'zipcode_98032', 'zipcode_98039', 'bathrooms', 'zipcode_98033', 'zipcode_98040', 'zipcode_98112', 'zipcode_98115', 'zipcode_98117', 'zipcode_98103', 'zipcode_98199', 'zipcode_98105', 'zipcode_98119', 'zipcode_98107', 'zipcode_98116', 'zipcode_98109', 'zipcode_98122', 'zipcode_98102', 'view_4_0', 'zipcode_98006', 'zipcode_98052', 'zipcode_98136', 'zipcode_98144', 'sqft_lot', 'zipcode_98008', 'zipcode_98005', 'grade_4', 'zipcode_98053', 'view_3_0', 'zipcode_98029', 'zipcode_98126', 'zipcode_98125', 'zipcode_98034', 'zipcode_98177', 'zipcode_98007', 'condition_5', 'zipcode_98074', 'zipcode_98075', 'grade_10', 'zipcode_98027', 'view_2_0', 'sqft_above', 'zipcode_98133', 'zipcode_98118', 'zipcode_98072', 'zipcode_98155', 'condition_4', 'condition_3', 'view_1_0', 'zipc

In [161]:
a=['grade_6', 'sqft_living', 'grade_7', 'grade_8', 'grade_5', 'zipcode_98004', 'zipcode_98023', 'grade_9', 'waterfront_1_0', 'sqft_living15', 'zipcode_98003', 'zipcode_98092', 'zipcode_98038', 'zipcode_98002', 'zipcode_98032', 'zipcode_98039', 'bathrooms', 'zipcode_98033', 'zipcode_98040', 'zipcode_98112', 'zipcode_98115', 'zipcode_98117', 'zipcode_98103', 'zipcode_98199', 'zipcode_98105', 'zipcode_98119', 'zipcode_98107', 'zipcode_98116', 'zipcode_98109', 'zipcode_98122', 'zipcode_98102', 'view_4_0', 'zipcode_98006', 'zipcode_98052', 'zipcode_98136', 'zipcode_98144', 'sqft_lot', 'zipcode_98008', 'zipcode_98005', 'grade_4', 'zipcode_98053', 'view_3_0', 'zipcode_98029', 'zipcode_98126', 'zipcode_98125', 'zipcode_98034', 'zipcode_98177', 'zipcode_98007', 'condition_5', 'zipcode_98074', 'zipcode_98075', 'grade_10', 'zipcode_98027', 'view_2_0', 'sqft_above', 'zipcode_98133', 'zipcode_98118', 'zipcode_98072', 'zipcode_98155', 'condition_4', 'condition_3', 'view_1_0', 'zipcode_98011', 'zipcode_98065', 'zipcode_98028', 'zipcode_98077', 'zipcode_98106', 'yr_renovated', 'zipcode_98108', 'zipcode_98059', 'zipcode_98056', 'zipcode_98024', 'grade_11', 'floors_3_0', 'zipcode_98166', 'zipcode_98146', 'zipcode_98045', 'zipcode_98019', 'zipcode_98014', 'zipcode_98070', 'zipcode_98010', 'zipcode_98058', 'zipcode_98178', 'zipcode_98055', 'sqft_basement', 'sqft_lot15', 'bedrooms_7', 'zipcode_98148', 'grade_13', 'bedrooms_2', 'condition_2', 'floors_2_0', 'bedrooms_6', 'zipcode_98188', 'zipcode_98168', 'bedrooms_5', 'bedrooms_9']

In [162]:
len(a)

97

In [163]:
b=['grade_7', 'sqft_living', 'zipcode_98004', 'waterfront_1_0', 'zipcode_98039', 'zipcode_98112', 'grade_13', 'grade_12', 'grade_11', 'grade_10', 'grade_6', 'zipcode_98040', 'grade_9', 'zipcode_98033', 'zipcode_98023', 'view_4_0', 'zipcode_98105', 'zipcode_98119', 'zipcode_98199', 'zipcode_98103', 'zipcode_98115', 'zipcode_98117', 'zipcode_98109', 'zipcode_98107', 'zipcode_98102', 'zipcode_98122', 'view_3_0', 'zipcode_98116', 'zipcode_98052', 'zipcode_98144', 'zipcode_98006', 'zipcode_98008', 'zipcode_98005', 'zipcode_98053', 'zipcode_98034', 'zipcode_98136', 'zipcode_98029', 'zipcode_98092', 'zipcode_98125', 'zipcode_98177', 'condition_5', 'zipcode_98007', 'zipcode_98126', 'zipcode_98074', 'zipcode_98075', 'zipcode_98027', 'zipcode_98133', 'zipcode_98118', 'view_2_0', 'zipcode_98072', 'zipcode_98155', 'sqft_above', 'view_1_0', 'yr_renovated', 'zipcode_98028', 'sqft_lot', 'zipcode_98106', 'zipcode_98011', 'condition_4', 'sqft_living15', 'bedrooms_3', 'zipcode_98108', 'floors_3_0', 'floors_2_0', 'bathrooms', 'zipcode_98146', 'zipcode_98056', 'zipcode_98077', 'zipcode_98024', 'zipcode_98065', 'zipcode_98059', 'zipcode_98045', 'zipcode_98019', 'bedrooms_2', 'zipcode_98014', 'zipcode_98166', 'bedrooms_7', 'zipcode_98168', 'floors_2_5', 'yr_built', 'bedrooms_9', 'zipcode_98038', 'zipcode_98055', 'zipcode_98058', 'bedrooms_10', 'zipcode_98010', 'zipcode_98148', 'sqft_lot15', 'zipcode_98178']

In [164]:
len(b)

89

In [165]:
print(a)

['grade_6', 'sqft_living', 'grade_7', 'grade_8', 'grade_5', 'zipcode_98004', 'zipcode_98023', 'grade_9', 'waterfront_1_0', 'sqft_living15', 'zipcode_98003', 'zipcode_98092', 'zipcode_98038', 'zipcode_98002', 'zipcode_98032', 'zipcode_98039', 'bathrooms', 'zipcode_98033', 'zipcode_98040', 'zipcode_98112', 'zipcode_98115', 'zipcode_98117', 'zipcode_98103', 'zipcode_98199', 'zipcode_98105', 'zipcode_98119', 'zipcode_98107', 'zipcode_98116', 'zipcode_98109', 'zipcode_98122', 'zipcode_98102', 'view_4_0', 'zipcode_98006', 'zipcode_98052', 'zipcode_98136', 'zipcode_98144', 'sqft_lot', 'zipcode_98008', 'zipcode_98005', 'grade_4', 'zipcode_98053', 'view_3_0', 'zipcode_98029', 'zipcode_98126', 'zipcode_98125', 'zipcode_98034', 'zipcode_98177', 'zipcode_98007', 'condition_5', 'zipcode_98074', 'zipcode_98075', 'grade_10', 'zipcode_98027', 'view_2_0', 'sqft_above', 'zipcode_98133', 'zipcode_98118', 'zipcode_98072', 'zipcode_98155', 'condition_4', 'condition_3', 'view_1_0', 'zipcode_98011', 'zipcode

In [166]:
len(x_cols)

112

In [167]:
len(x_cols)

112

## Model Refinement I

In [168]:
outcome = 'price'
x_cols = ['grade_7', 'sqft_living15', 'sqft_living', 'grade_6', 'zipcode_98023', 'zipcode_98004', 'grade_8', 'grade_5', 'grade_9', 'waterfront_1_0', 'zipcode_98003', 'zipcode_98092', 'zipcode_98038', 'zipcode_98002', 'zipcode_98032', 'zipcode_98039', 'bathrooms', 'zipcode_98033', 'zipcode_98040', 'zipcode_98112', 'zipcode_98115', 'zipcode_98117', 'zipcode_98103', 'zipcode_98199', 'zipcode_98105', 'zipcode_98119', 'zipcode_98107', 'zipcode_98116', 'zipcode_98109', 'zipcode_98122', 'zipcode_98102', 'view_4_0', 'zipcode_98006', 'zipcode_98052', 'zipcode_98136', 'zipcode_98144', 'sqft_lot', 'zipcode_98008', 'zipcode_98005', 'grade_4', 'zipcode_98053', 'view_3_0', 'zipcode_98029', 'zipcode_98126', 'zipcode_98125', 'zipcode_98034', 'zipcode_98177', 'zipcode_98007', 'condition_5', 'zipcode_98074', 'zipcode_98075', 'grade_10', 'zipcode_98027', 'view_2_0', 'sqft_above', 'zipcode_98133', 'zipcode_98118', 'zipcode_98072', 'zipcode_98155', 'condition_4', 'condition_3', 'view_1_0', 'zipcode_98011', 'zipcode_98065', 'zipcode_98028', 'zipcode_98077', 'zipcode_98106', 'yr_renovated', 'zipcode_98108', 'zipcode_98059', 'zipcode_98056', 'zipcode_98024', 'grade_11', 'floors_3_0', 'zipcode_98166', 'zipcode_98146', 'zipcode_98045', 'zipcode_98019', 'zipcode_98014', 'zipcode_98070', 'zipcode_98010', 'zipcode_98058', 'zipcode_98178', 'zipcode_98055', 'sqft_basement', 'sqft_lot15', 'bedrooms_7', 'zipcode_98148', 'grade_13', 'bedrooms_2', 'condition_2', 'floors_2_0', 'bedrooms_6', 'zipcode_98188', 'zipcode_98168', 'bedrooms_5', 'bedrooms_9']

In [169]:
# Fitting the actual model
predictors = '+'.join(x_cols)
formula = outcome + '~' + predictors
model = ols(formula=formula, data=df_feats).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.885
Model:                            OLS   Adj. R-squared:                  0.885
Method:                 Least Squares   F-statistic:                     1706.
Date:                Tue, 28 Apr 2020   Prob (F-statistic):               0.00
Time:                        17:07:04   Log-Likelihood:                 6568.0
No. Observations:               21594   AIC:                        -1.294e+04
Df Residuals:                   21496   BIC:                        -1.216e+04
Df Model:                          97                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         12.7573      0.040    320.698      0.000      12.679      12.835
grade_7           -0.5247      0.021    -25.489      0.000      -0.565      -0.484
sqft_living15      0.0493      0.002     22.558      0.000       0.045       0.054
sqft_living        0.0968      0.006     16.564      0.000       0.085       0.108
grade_6           -0.5922      0.021    -27.602      0.000      -0.634      -0.550
zipcode_98023     -0.0718      0.009     -7.951      0.000      -0.090      -0.054
zipcode_98004      1.0765      0.011     96.810      0.000       1.055       1.098
grade_8           -0.4433      0.020    -22.066      0.000      -0.483      -0.404
grade_5           -0.6571      0.024    -26.938      0.000      -0.705      -0.609
grade_9           -0.3286      0.020    -16.532      0.000      -0.368      -0.290
waterfront_1_0     0.4449      0.018     24.117      0.000       0.409       0.481
zipcode_98003     -0.0279      0.011     -2.436      0.015      -0.050      -0.005
zipcode_98092     -0.0315      0.010     -3.020      0.003      -0.052      -0.011
zipcode_98038      0.1168      0.009     13.731      0.000       0.100       0.133
zipcode_98002     -0.0513      0.013     -3.822      0.000      -0.078      -0.025
zipcode_98032     -0.0630      0.017     -3.807      0.000      -0.095      -0.031
zipcode_98039      1.2398      0.026     47.703      0.000       1.189       1.291
bathrooms          0.0281      0.002     12.820      0.000       0.024       0.032
zipcode_98033      0.7401      0.010     77.008      0.000       0.721       0.759
zipcode_98040      0.8293      0.012     70.294      0.000       0.806       0.852
zipcode_98112      1.0367      0.012     86.053      0.000       1.013       1.060
zipcode_98115      0.8075      0.009     92.485      0.000       0.790       0.825
zipcode_98117      0.8135      0.009     90.753      0.000       0.796       0.831
zipcode_98103      0.8372      0.009     93.661      0.000       0.820       0.855
zipcode_98199      0.8412      0.011     75.113      0.000       0.819       0.863
zipcode_98105      0.9356      0.013     73.003      0.000       0.910       0.961
zipcode_98119      0.9767      0.014     68.973      0.000       0.949       1.004
zipcode_98107      0.8608      0.012     70.922      0.000       0.837       0.885
zipcode_98116      0.7517      0.011     68.488      0.000       0.730       0.773
zipcode_98109      0.9944      0.018     55.583      0.000       0.959       1.029
zipcode_98122      0.8106      0.012     69.680      0.000       0.788       0.833
zipcode_98102      0.9642      0.018     52.446      0.000       0.928       1.000
view_4_0           0.2925      0.013     22.869      0.000       0.267       0.318
zipcode_98006      0.5929      0.009     63.570      0.000       0.575       0.611
zipcode_9805

In [170]:
X = df_feats[x_cols]

In [171]:
X.head()

grade_7  sqft_living15  sqft_living  grade_6  zipcode_98023  zipcode_98004  \
0        1      -1.035409    -1.125683        0              0              0   
1        1      -0.326807     0.709834        0              0              0   
2        0       1.126666    -2.131831        1              0              0   
3        1      -0.990173     0.070792        0              0              0   
4        0      -0.134239    -0.292723        0              0              0   

   grade_8  grade_5  grade_9  waterfront_1_0  ...  zipcode_98148  grade_13  \
0        0        0        0               0  ...              0         0   
1        0        0        0               0  ...              0         0   
2        0        0        0               0  ...              0         0   
3        0        0        0               0  ...              0         0   
4        1        0        0               0  ...              0         0   

   bedrooms_2  condition_2  floors_2_0  bedrooms_6  zipcode_98188  \
0           0            0           0           0              0   
1           0            0           1           0              0   
2           1            0           0           0              0   
3           0            0           0           0              0   
4           0            0           0           0              0   

   zipcode_98168  bedrooms_5  bedrooms_9  
0              0           0           0  
1              0           0           0  
2              0           0           0  
3              0           0           0  
4              0           0           0  

[5 rows x 97 columns]

In [172]:
fig = sm.graphics.qqplot(model.resid, dist=stats.norm, line='45', fit=True)

In [173]:
plt.scatter(model.predict(df_feats[x_cols]), model.resid)
plt.plot(model.predict(df_feats[x_cols]), [0 for i in range(len(df_feats))])

In [174]:
vif = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
list(zip(x_cols, vif))

[('grade_7', 86.78498522120823),
 ('sqft_living15', 3.210888488418228),
 ('sqft_living', 23.016517311781026),
 ('grade_6', 21.55868504532976),
 ('zipcode_98023', 1.2655115162110668),
 ('zipcode_98004', 1.216143216071013),
 ('grade_8', 56.23484952333962),
 ('grade_5', 3.4030894577343616),
 ('grade_9', 24.109533828675897),
 ('waterfront_1_0', 1.5518794091058816),
 ('zipcode_98003', 1.1474959183039652),
 ('zipcode_98092', 1.1916613047719953),
 ('zipcode_98038', 1.3290484346719755),
 ('zipcode_98002', 1.1196574947487616),
 ('zipcode_98032', 1.069170107530686),
 ('zipcode_98039', 1.052427518452627),
 ('bathrooms', 3.2238069068262343),
 ('zipcode_98033', 1.2400626901956693),
 ('zipcode_98040', 1.2213225543383506),
 ('zipcode_98112', 1.2118716678670949),
 ('zipcode_98115', 1.3798709031587486),
 ('zipcode_98117', 1.3781122996343032),
 ('zipcode_98103', 1.491839635137205),
 ('zipcode_98199', 1.2365972742914628),
 ('zipcode_98105', 1.171037158786296),
 ('zipcode_98119', 1.1490478943763824),
 ('z

## Cross Validation

In [175]:
linreg = LinearRegression()
cvs_10=cross_val_score(linreg, X, y, cv=10)
cvs_10_res_mean = round(np.mean(cvs_10),5)
cv_10_res_std = round(np.std(cvs_10),5)
print(f'Mean cross validation score for 10 fold is {cvs_10_res_mean}')
print(f'Standard deviation of the score is {cv_10_res_std}')


Mean cross validation score for 10 fold is 0.88227
Standard deviation of the score is 0.00346


##  Location! Location! Location!

In [176]:
income_df=pd.read_csv('kc_house_data.csv')

In [177]:
from uszipcode import SearchEngine, Zipcode
fdfdrerrerererreeeerereere4y444w444444rryyydyhdyddt7e9[djci9dxxxzxc8phc8ccp                    hchhdhcjcjjjcjcjcjcjcjc           V]

SyntaxError: invalid syntax (<ipython-input-177-f0716acf6ba6>, line 2)

In [ ]:
search = SearchEngine(simple_zipcode=False)

In [ ]:
def zco(x):
    income = search.by_zipcode(x)
    
    return income.median_household_income if income else x
   

In [ ]:
income_df['income'] = income_df['zipcode'].astype(int).astype(str).apply(zco)

In [ ]:
income_df.head()

In [ ]:
# income_df=pd.read_csv('kc_house_data.csv')

In [ ]:
fig = plt.figure(figsize = (20,7))
fig = sns.boxplot(x='zipcode', y='price', data=income_df, showfliers = False)
plt.title('Prices of Houses Sold in King County by Zip Code')
plt.xlabel("Zip Code")
plt.xticks(rotation=90)
plt.ylabel('Sale Price, $')
plt.show()


https://www.zillow.com/bellevue-wa-98004/

https://www.zillow.com/mercer-island-wa-98040/

https://www.zillow.com/medina-wa-98039/

https://www.zillow.com/seattle-wa-98112/

In [ ]:
high_price_zipcodes=[98040,98004,98039,98112]
for i in high_price_zipcodes:
    zipcode = search.by_zipcode(i)
    income=zipcode.median_household_income
    print('Zip code:',i ,'-------   Median household income of all the houses in this zipcode:' ,income)